docs/design/2021-10-15-analyze-predicate-columns.md

```diff
@@ -0,0 +1,73 @@
+
+## Proposal
+
+We want to support the `ANALYZE` option to only collect the statistics of predicate columns and indexed columns. Sometimes predicate columns may vary since new queries come in continuously, so we want another option to only collect the statistics of user-specified columns and indexed columns. 
+
+### ANALYZE Syntax
+
+```
+ANALYZE TABLE TableNameList (ALL COLUMNS | PREDICATE COLUMNS | COLUMNS ColumnNameList)? AnalyzeOptionListOpt
+ANALYZE TABLE TableName (PARTITION PartitionNameList)? (PREDICATE COLUMNS | COLUMNS ColumnNameList)? AnalyzeOptionListOpt
+```
+1. FOR `ANALYZE ALL COLUMNS`, the statistics of all columns would be collected.
+1. For `ANALYZE PREDICATE COLUMNS`, the statistics of predicate columns and indexed columns would be collected. If there is no predicate columns in record, TiDB would collect the statistics of all columns and give a warning to the client.
+2. For `ANALYZE COLUMNS ColumnNameList`, the statistics of the columns in `ColumnNameList` and indexed columns would be collected.
+3. By default, the statistics of all columns would be collected.
+
+### Track Predicate Columns
+
+We build a new system table `mysql.column_stats_usage` to track predicate columns.
+```sql
+CREATE TABLE IF NOT EXISTS mysql.column_stats_usage (
+	table_id BIGINT(64) NOT NULL,
+	column_id BIGINT(64) NOT NULL,
+	last_used_at TIMESTAMP,
+	last_analyzed_at TIMESTAMP,
+	PRIMARY KEY (table_id, column_id) CLUSTERED
+);
+```
+`table_id` and `column_id` are the IDs of the table and the column respectively. `last_used_at` is the timestamp at which the column stats were used(needed). `last_analyzed_at` is the timestamp at which the column stats are updated.
+
+During the plan optimization phase, we track which column stats are used(needed) in memory. `updateStatsWorker` periodically dumps column stats usage(in memory) into `mysql.column_stats_usage`(in disk). Since tracking predicate columns in memory is a high frequency operation, we provide a switch to control whether to enable tracking predicate columns for insurance.
+
+Each time when executing `ANALYZE PREDICATE COLUMNS`, we read predicate columns(the columns whose `last_used_at` is not null) from `mysql.column_stats_usage` and then decide which columns' statistics need to be collected.
+
+If a table or a column is deleted, the related rows in `mysql.column_stats_usage` also need to be deleted.
+
+Besides, we provide the syntax `show column_stats_usage` to let users see which columns' stats are used(needed) or updated.
+
+### Auto Analyze Logic
+
+By default, auto-analyze would collect the statistics of all columns. We can modify the `ANALYZE` configuration for a specific table. The configuration includes the column option(all columns/predicate columns/user-specified columns). Once we set the configuration for the table, auto-analyze can decide which columns' statistics need to be collected according to the column option.
```

Comment:
How to modify the config for auto-analyze? Is there any new syntax?

Reply:
#29056 provides syntax to modify a certain table's analyze config, which is used by auto-analyze.